src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SpecializedTasks.cs

```diff
@@ -86,6 +87,88 @@ public static ValueTask<T[]> WhenAll<T>(IEnumerable<ValueTask<T>> tasks)
+            Func<TArg, CancellationToken, ValueTask<TIntermediate>> func,
+            Func<TIntermediate, TArg, TResult> transform,
+            TArg arg,
+            CancellationToken cancellationToken)
+        {
+            if (func is null)
+                throw new ArgumentNullException(nameof(func));
+            if (transform is null)
+                throw new ArgumentNullException(nameof(transform));
+
+            var intermediateResult = func(arg, cancellationToken);
+            if (intermediateResult.IsCompletedSuccessfully)
+            {
+                // Synchronous fast path if 'func' completes synchronously
+                var result = intermediateResult.Result;
+                if (cancellationToken.IsCancellationRequested)
+                    return new ValueTask<TResult>(Task.FromCanceled<TResult>(cancellationToken));
+
+                return new ValueTask<TResult>(transform(result, arg));
+            }
+            else if (intermediateResult.IsCanceled && cancellationToken.IsCancellationRequested)
+            {
+                // Synchronous fast path if 'func' cancels synchronously
+                return new ValueTask<TResult>(Task.FromCanceled<TResult>(cancellationToken));
+            }
+            else
+            {
+                // Asynchronous fallback path
+                return UnwrapAndTransformAsync(intermediateResult, transform, arg, cancellationToken);
+            }
+
+            static ValueTask<TResult> UnwrapAndTransformAsync(ValueTask<TIntermediate> intermediateResult, Func<TIntermediate, TArg, TResult> transform, TArg arg, CancellationToken cancellationToken)
+            {
+                // Apply the transformation function once a result is available. The behavior depends on the final
+                // status of 'intermediateResult' and the 'cancellationToken'.
+                //
+                // | 'intermediateResult'       | 'cancellationToken' | Behavior                                 |
+                // | -------------------------- | ------------------- | ---------------------------------------- |
+                // | Ran to completion          | Not cancelled       | Apply transform                          |
+                // | Ran to completion          | Cancelled           | Cancel result without applying transform |
+                // | Cancelled (matching token) | Cancelled           | Cancel result without applying transform |
+                // | Cancelled (mismatch token) | Not cancelled       | Cancel result without applying transform |
+                // | Cancelled (mismatch token) | Cancelled           | Cancel result without applying transform |
+                // | Direct fault¹              | Not cancelled       | Fault result without applying transform  |
+                // | Direct fault¹              | Cancelled           | Cancel result without applying transform |
```

Comment:
this seems fine.  but it feels a bit lossy.  is cancellation preferred here over bubbling the fault up?

Reply:
Passing through the fault is heavily preferred, but we can't do that without adding a performance cost to the common cases. See https://github.com/dotnet/roslyn/pull/55138#discussion_r677725554.

Reply:
❗ This documentation is not correct. These are the results for _indirect_ faults. I will fix.

Reply:
OK now the table is correct for these two. The unfortunate case is the final case (indirect fault combined with a cancellation request).